src/process.cr

```diff
@@ -117,14 +117,31 @@ class Process
   # *run_hooks* should ALWAYS be `true` unless `exec` is used immediately after fork.
   # Channels, `IO` and other will not work reliably if *run_hooks* is `false`.
   protected def self.fork_internal(run_hooks : Bool = true)
-    pid = LibC.fork
-    case pid
+    newmask = uninitialized LibC::SigsetT
+    oldmask = uninitialized LibC::SigsetT
+
+    LibC.sigfillset(pointerof(newmask))
+    ret = LibC.pthread_sigmask(LibC::SIG_SETMASK, pointerof(newmask), pointerof(oldmask))
+    raise Errno.new("Failed to disable signals") unless ret == 0
+
+    case pid = LibC.fork
     when 0
+      # child:
       pid = nil
-      Process.after_fork_child_callbacks.each(&.call) if run_hooks
+      if run_hooks
+        Process.after_fork_child_callbacks.each(&.call)
+        LibC.pthread_sigmask(LibC::SIG_SETMASK, pointerof(oldmask), nil)
+      end
```

Comment:
In chose to not reenable signals in a fork/exec situation. I just leave the handlers and mask in place, and let `execve(2)` override that for the new process.

Reply:
I think if you don't reenable the signals, they will remain blocked across the exec. So, it is necessary to reenable the signals.

```markdown
**The new process shall inherit at least the following attributes from the calling process image**:

[XSI] [Option Start] Nice value (see nice()) [Option End]

[XSI] [Option Start] semadj values (see semop()) [Option End]

Process ID

Parent process ID

Process group ID

Session membership

Real user ID

Real group ID

Supplementary group IDs

Time left until an alarm clock signal (see alarm())

Current working directory

Root directory

File mode creation mask (see umask())

[XSI] [Option Start] File size limit (see ulimit()) [Option End]

**Process signal mask (see sigprocmask())**

Pending signal (see sigpending())

tms_utime, tms_stime, tms_cutime, and tms_cstime (see times())

[XSI] [Option Start] Resource limits [Option End]

[XSI] [Option Start] Controlling terminal [Option End]

[XSI] [Option Start] Interval timers [Option End]

The initial thread of the new process shall inherit at least the following attributes from the calling thread:

Signal mask (see sigprocmask() and pthread_sigmask())

Pending signals (see sigpending())
```

Reference:
http://pubs.opengroup.org/onlinepubs/009695399/functions/exec.html